File: cpp/shoe_sorting_demo/planning.cpp
#include "planning.h"

#include <cmath>
#include <iostream>
#include <map>
#include <stdexcept>
#include <tuple>
#include <vector>

#include "pairing.h"
#include "visualize.h"

namespace shoe_sorting {

namespace {
double QuatXAxisYaw(const robot_sdk::Orientation& q) {
  double rxx = 1.0 - 2.0 * q.y * q.y - 2.0 * q.z * q.z;
  double ryx = 2.0 * q.x * q.y + 2.0 * q.w * q.z;
  return std::atan2(ryx, rxx);
}
}  // namespace

static auto BboxKey(const std::vector<int>& b) {
  return std::make_tuple(b[0], b[1], b[2], b[3]);
}

std::vector<std::vector<int>> HeadImageRefBboxes(
    const std::vector<robot_sdk::ShoesIntentStep>& steps) {
  using Key = std::tuple<int, int, int, int>;
  std::map<Key, Key> to_to_from;
  for (const auto& m : steps) to_to_from[BboxKey(m.to_bbox)] = BboxKey(m.from_bbox);
  std::vector<std::vector<int>> out;
  for (const auto& m : steps) {
    Key cur = BboxKey(m.from_bbox);
    for (size_t i = 0; i <= steps.size(); ++i) {
      auto it = to_to_from.find(cur);
      if (it == to_to_from.end()) break;
      cur = it->second;
    }
    out.push_back({std::get<0>(cur), std::get<1>(cur), std::get<2>(cur), std::get<3>(cur)});
  }
  return out;
}

std::vector<robot_sdk::ShoesIntentStep> GetShoeMoveSteps(
    const robot_sdk::ImageQueryNotifyData& head_image,
    std::vector<DetectedShoe>& shoes,
    const std::vector<std::vector<double>>& carrier_box,
    const std::vector<std::vector<double>>& front_edge,
    const std::string& vis_prefix,
    bool move_all) {
  const auto& image_data = head_image.rgb_image.data;
  bool save_vis = !vis_prefix.empty();

  // 1. 配对
  PairShoes(image_data, shoes);
  int paired = 0;
  for (const auto& s : shoes) if (s.pair_id > 0) ++paired;
  std::cout << "  配对: " << paired << "/" << shoes.size() << " 已配对"
            << (shoes.size() - paired) << " 未配对" << std::endl;

  if (save_vis) WriteRgbPng(vis_prefix + "_pairs.png", DrawPairs(DecodeImage(image_data), shoes));

  // 2. 放置区投影
  robot_sdk::ProjectPlaceLineAndPlaneRequest proj_req;
  {
    robot_sdk::ProjectPlaceLineAndPlaneView view;
    view.rgb_image = head_image.rgb_image;
    view.depth_image = head_image.depth_image;
    view.tf_goal = head_image.tf_goal;
    view.camera_info_k = head_image.camera_info_k;
    proj_req.views.push_back(std::move(view));
    for (const auto& pt : carrier_box)
      proj_req.carrier_box.push_back({static_cast<float>(pt[0]), static_cast<float>(pt[1]), static_cast<float>(pt[2])});
    for (const auto& pt : front_edge)
      proj_req.front_edge.push_back({static_cast<float>(pt[0]), static_cast<float>(pt[1]), static_cast<float>(pt[2])});
  }
  auto proj_resp = robot_sdk::eco_ProjectPlaceLineAndPlane(proj_req);
  if (!proj_resp || proj_resp->code != 0 || proj_resp->data.empty())
    throw std::runtime_error("投影摆放区失败");
  const auto& place_line = proj_resp->data[0].place_line;
  const auto& place_plane = proj_resp->data[0].place_plane;
  const auto& place_yaw = proj_resp->data[0].place_yaw;

  cv::Mat placement_vis;
  if (save_vis) {
    placement_vis = DrawPlacement(DecodeImage(image_data), place_line, place_plane);
    WriteRgbPng(vis_prefix + "_placement.png", placement_vis);
  }

  // 3. 步进规划
  robot_sdk::ShoesIntentPlanRequest req;
  for (const auto& shoe : shoes) {
    robot_sdk::ShoesIntentShoeItem item;
    item.bbox = shoe.bbox; item.pair_id = shoe.pair_id;
    req.shoes.push_back(item);
  }

  if (!move_all) {
    static const double kTiltThreshRad = 30.0 * 3.14159265358979323846 / 180.0;
    for (auto& item : req.shoes) {
      item.is_tilted = true;
      try {
        robot_sdk::BboxItem fb;
        fb.id = 0;
        fb.left_up_x = static_cast<float>(item.bbox[0]);
        fb.left_up_y = static_cast<float>(item.bbox[1]);
        fb.right_down_x = static_cast<float>(item.bbox[2]);
        fb.right_down_y = static_cast<float>(item.bbox[3]);
        robot_sdk::GetPoseData gp;
        gp.bbox = {fb}; gp.rgb_image = head_image.rgb_image;
        gp.depth_image = head_image.depth_image; gp.tf_map = head_image.tf_goal;
        robot_sdk::GetPoseNotifyData pr;
        if (robot_sdk::eco_GetPose(gp, pr, 6000).code == 0 && !pr.pose_results.empty()) {
          double diff = std::fmod(std::abs(QuatXAxisYaw(pr.pose_results[0].orientation) - place_yaw), 3.14159265358979323846);
          item.is_tilted = std::min(diff, 3.14159265358979323846 - diff) >= kTiltThreshRad;
        }
      } catch (const std::exception&) {}
    }
  }

  req.place_line = place_line;
  req.place_plane = place_plane;
  auto plan_result = robot_sdk::eco_ShoesIntentPlan(req, move_all);
  if (!plan_result || plan_result->code != 0)
    throw std::runtime_error("ShoesIntentPlan 失败: code=" + std::to_string(plan_result ? plan_result->code : -1));
  std::vector<robot_sdk::ShoesIntentStep> steps = plan_result->steps;
  std::cout << "  规划: " << steps.size() << " 步" << std::endl;

  if (save_vis) {
    cv::Mat base = placement_vis.empty() ? DecodeImage(image_data) : placement_vis;
    WriteRgbPng(vis_prefix + "_plan.png", DrawPlans(base, steps));
  }
  return steps;
}

}  // namespace shoe_sorting

File: python/shoe_sorting_demo/planning.py
from __future__ import annotations

import math,sys
from typing import Dict, List, Sequence, Tuple
from pathlib import Path

from bajie_sdk import (
    BajieRobot,
    MovePlanStep,
    ObjectDetection,
    PairedBBox,
    Quatf,
    RGBDViewWithPose,
)
try:
    from .pairing import pair_shoes
    from .visualize_plans import (draw_pairs_on_rgb,draw_placement_on_rgb,draw_plans_on_rgb,write_rgb_png)
except ImportError:
    if (_d := str(Path(__file__).resolve().parent)) not in sys.path: sys.path.insert(0, _d)
    from pairing import pair_shoes
    from visualize_plans import (draw_pairs_on_rgb,draw_placement_on_rgb,draw_plans_on_rgb,write_rgb_png)


def _bbox_key(bbox: Sequence[int]) -> tuple[int, int, int, int]:
    return (int(bbox[0]), int(bbox[1]), int(bbox[2]), int(bbox[3]))


def head_image_ref_bboxes_for_move_steps(
    move_steps: Sequence[MovePlanStep],
) -> List[Tuple[int, int, int, int]]:
    """为每一步规划给出头图上的**参考框**（与当步 ``from_bbox`` 同尺、但取「链首」位置）.

    规划会按模拟摆放改写每步 ``from_bbox``，而执行时头图不刷新；``eco_match_obj_views`` 需在**首帧头图**
    上裁切。只根据 ``move_steps`` 判断**是否连续移动**（本步 ``from_bbox`` 与上一步 ``to_bbox`` 四元组相同）：

    * 新链的第一步：该步 ``from_bbox`` 即该物体在头图中的**原本**位置；
    * 连续步：建 ``{ to_bbox: from_bbox }``（每步一条）；对当前步的 ``from``，若在字典中作为 ``to`` 出现则
      ``cur = d[cur]`` 循环直到找不到，``cur`` 即链头 ``from``。
    """
    if not move_steps:
        return []
    steps = sorted(move_steps, key=lambda s: s.step_id)
    to_to_from: Dict[Tuple[int, int, int, int], Tuple[int, int, int, int]] = {
        _bbox_key(m.to_bbox): _bbox_key(m.from_bbox) for m in steps
    }
    n = len(steps)
    out: List[Tuple[int, int, int, int]] = []
    for m in steps:
        cur = _bbox_key(m.from_bbox)
        for _ in range(n + 1):
            if cur not in to_to_from:
                break
            cur = to_to_from[cur]
        else:
            raise ValueError("to→from 出现环，无法归约到链头")
        out.append(cur)
    return out


def _quat_x_axis_yaw(q: Quatf) -> float:
    """从四元数提取局部 x 轴在地平面投影的 yaw（弧度）。"""
    x, y, z, w = float(q.x), float(q.y), float(q.z), float(q.w)
    rxx = 1.0 - 2.0 * y * y - 2.0 * z * z
    ryx = 2.0 * x * y + 2.0 * w * z
    return math.atan2(ryx, rxx)


def get_shoe_move_steps(
    robot: BajieRobot, head_image: RGBDViewWithPose,
    shoe_items: Sequence[ObjectDetection], *,
    carrier_box: Sequence[Sequence[float]],
    front_edge: Sequence[Sequence[float]],
    move_all: bool = True, vis_prefix: str | None = None,
    tilt_thresh_deg: float = 30.0,
) -> List[MovePlanStep]:
    """配对/线面投影/步进规划。move_all=False 时仅整理摆歪(is_tilted)的鞋子。"""
    pair_map = pair_shoes(robot, head_image, shoe_items)

    from bajie_sdk.vision_http import project_place_line_and_plane
    proj = project_place_line_and_plane(head_image, carrier_box, front_edge)
    if not proj: raise ValueError("投影失败：放置区不在视野范围内")
    row = proj[0]

    paired_bbox = [PairedBBox(
        bbox=(int(it.bbox[0]), int(it.bbox[1]), int(it.bbox[2]), int(it.bbox[3])),
        pair_id=pair_map.get(str(it.uuid), 0),
    ) for it in shoe_items]

    if not move_all:
        thresh = math.radians(tilt_thresh_deg)
        for pb in paired_bbox:
            pb.is_tilted = True
            try:
                pose = robot.eco_computeObjectPose([pb.bbox], head_image, timeout_sec=6.0)[0]
                diff = abs(_quat_x_axis_yaw(pose.orientation) - row.place_yaw) % math.pi
                pb.is_tilted = min(diff, math.pi - diff) >= thresh
            except Exception:
                pass

    rgb, plc_rgb = head_image.rgb_image.img, None
    base = str(vis_prefix).rstrip() if vis_prefix and str(vis_prefix).strip() else None
    if base:
        write_rgb_png(f"{base}_pairs.png",  draw_pairs_on_rgb(rgb, shoe_items, pair_map))
        plc_rgb = draw_placement_on_rgb(rgb, row.place_line, row.tmp_palce_area)
        write_rgb_png(f"{base}_placement.png", plc_rgb)

    plan = robot.eco_shoe_placement_plan(
        head_image, carrier_box, front_edge, paired_bbox,
        move_all=move_all, timeout=20.0)
    steps = plan[0] if isinstance(plan, tuple) else plan
    if not steps: return []

    if plc_rgb is not None:
        write_rgb_png(f"{base}_plan.png", draw_plans_on_rgb(plc_rgb, steps))
    return list(steps)
